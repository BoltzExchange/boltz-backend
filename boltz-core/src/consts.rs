pub const ECDSA_BYTES_TO_GRIND: usize = 2;

pub const PREIMAGE_DUMMY: [u8; 0] = [];

pub const STUB_SCHNORR_SIGNATURE_LENGTH: usize = 64;
