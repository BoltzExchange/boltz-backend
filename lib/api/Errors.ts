export default {
  COULD_NOT_PARSE_HEX: (argName: string): string =>
    `could not parse hex string: ${argName}`,
  INVALID_PARAMETER: (argName: string): string =>
    `invalid parameter: ${argName}`,
  UNDEFINED_PARAMETER: (argName: string): string =>
    `undefined parameter: ${argName}`,
  UNSUPPORTED_PARAMETER: (symbol: string, argName: string): string =>
    `${symbol} does not support ${argName}`,
  INVALID_SWAP_STATUS: (status: string): string =>
    `invalid swap status: ${status}`,
};
