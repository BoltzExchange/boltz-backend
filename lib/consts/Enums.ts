export enum ErrorCodePrefix {
  General = 0,
  Grpc = 1,
  Service = 2,
  Wallet = 3,
  Chain = 4,
  Lnd = 5,
  Swap = 6,
  Rates = 7,
  Backup = 8,
}

export enum ClientStatus {
  Disconnected,
  Connected,
  OutOfSync,
}

export enum OrderSide {
  BUY,
  SELL,
}

export enum SwapUpdateEvent {
  SwapCreated = 'swap.created',
  SwapExpired = 'swap.expired',

  ChannelCreated = 'channel.created',

  InvoiceSet = 'invoice.set',
  InvoicePaid = 'invoice.paid',
  InvoicePending = 'invoice.pending',
  InvoiceSettled = 'invoice.settled',
  InvoiceFailedToPay = 'invoice.failedToPay',

  TransactionFailed = 'transaction.failed',
  TransactionMempool = 'transaction.mempool',
  TransactionClaimed = 'transaction.claimed',
  TransactionRefunded = 'transaction.refunded',
  TransactionConfirmed = 'transaction.confirmed',

  // Events for the prepay miner fee Reverse Swap protocol
  MinerFeePaid = 'minerfee.paid',
}

export enum ChannelCreationType {
  Auto = 'auto',
  Create = 'create',
}

export enum ChannelCreationStatus {
  Attempted = 'attempted',
  Created = 'created',
  Settled = 'settled',
  Abandoned = 'abandoned',
}

export enum ServiceWarning {
  ReverseSwapsDisabled = 'reverse.swaps.disabled',
}

export enum ServiceInfo {
  PrepayMinerFee = 'prepay.minerfee',
}

export enum Network {
  Mainnet = 'mainnet',
  Testnet = 'testnet',
  Simnet = 'simnet',
  Regtest = 'regtest',
}

export enum SwapType {
  Submarine = 'submarine',
  ReverseSubmarine = 'reversesubmarine',
}
