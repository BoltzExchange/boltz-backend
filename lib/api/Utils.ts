import { Request, Response } from 'express';
import Logger from '../Logger';
import { getHexBuffer } from '../Utils';
import MarkedSwapRepository from '../db/repositories/MarkedSwapRepository';
import CountryCodes from '../service/CountryCodes';
import ServiceErrors from '../service/Errors';
import Errors from './Errors';

type ApiArgument = {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'object';
  hex?: boolean;
  optional?: boolean;
};

// Some endpoints are getting spammed on mainnet, so we don't log the warnings for them
const errorsNotToLog: any[] = [
  ServiceErrors.SWAP_NO_LOCKUP().message,
  ServiceErrors.ETHEREUM_NOT_ENABLED().message,
];

/**
 * Validates that all required arguments were provided in the body correctly
 *
 * @returns the validated arguments
 */
export const validateRequest = (
  body: Record<string, any>,
  argsToCheck: ApiArgument[],
) => {
  const response: any = {};

  argsToCheck.forEach((arg) => {
    const value = body[arg.name];

    if (value !== undefined) {
      if (typeof value === arg.type) {
        if (arg.hex && value !== '') {
          const buffer = getHexBuffer(value);

          if (buffer.length === 0) {
            throw Errors.COULD_NOT_PARSE_HEX(arg.name);
          }

          response[arg.name] = buffer;
        } else {
          response[arg.name] = value;
        }
      } else {
        throw Errors.INVALID_PARAMETER(arg.name);
      }
    } else if (!arg.optional) {
      throw Errors.UNDEFINED_PARAMETER(arg.name);
    }
  });

  return response;
};

export const errorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  error: unknown,
  statusCode = 400,
): void => {
  if (typeof error === 'string') {
    writeErrorResponse(logger, req, res, statusCode, { error });
  } else {
    const errorObject = error as any;

    // Bitcoin Core related errors
    if (errorObject.details) {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.details,
      });
      // Custom error when broadcasting a refund transaction fails because
      // the locktime requirement has not been met yet
    } else if (errorObject.timeoutBlockHeight) {
      writeErrorResponse(logger, req, res, statusCode, error);
      // Everything else
    } else {
      writeErrorResponse(logger, req, res, statusCode, {
        error: errorObject.message,
      });
    }
  }
};

export const successResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(200);

  if (typeof data === 'object') {
    res.json(data);
  } else {
    res.write(data);
    res.end();
  }
};

export const createdResponse = (res: Response, data: unknown) => {
  setContentTypeJson(res);
  res.status(201).json(data);
};

export const writeErrorResponse = (
  logger: Logger,
  req: Request,
  res: Response,
  statusCode: number,
  error: any,
) => {
  if (!errorsNotToLog.includes(error?.error || error)) {
    logger.warn(
      `Request ${req.method} ${req.originalUrl} ${
        req.body && Object.keys(req.body).length > 0
          ? `${JSON.stringify(req.body)} `
          : ''
      }failed: ${JSON.stringify(error)}`,
    );
  }

  setContentTypeJson(res);
  res.status(statusCode).json(error);
};

export const setContentTypeJson = (res: Response) => {
  res.set('Content-Type', 'application/json');
};

export const checkPreimageHashLength = (preimageHash: Buffer) => {
  if (preimageHash.length !== 32) {
    throw `invalid preimage hash length: ${preimageHash.length}`;
  }
};

export const markSwap = async (
  countryCodes: CountryCodes,
  ip: string,
  swapId: string,
) => {
  if (!countryCodes.isRelevantCountry(countryCodes.getCountryCode(ip))) {
    return;
  }

  await MarkedSwapRepository.addMarkedSwap(swapId);
};
