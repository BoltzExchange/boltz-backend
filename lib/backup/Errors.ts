import { concatErrorCode } from '../Utils';
import { ErrorCodePrefix } from '../consts/Enums';
import { Error } from '../consts/Types';

export default {
  BACKUP_DISABLED: (): Error => ({
    message: 'backups are disabled because of incomplete configuration',
    code: concatErrorCode(ErrorCodePrefix.Backup, 0),
  }),
};
