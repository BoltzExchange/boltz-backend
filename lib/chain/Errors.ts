import { concatErrorCode } from '../Utils';
import { ErrorCodePrefix } from '../consts/Enums';
import { Error } from '../consts/Types';

export default {
  NO_RAWTX: (): Error => ({
    message: 'pubrawtx ZMQ notifications are not enabled',
    code: concatErrorCode(ErrorCodePrefix.Chain, 1),
  }),
  NO_BLOCK_NOTIFICATIONS: (): Error => ({
    message: 'no ZMQ block notifications are enabled',
    code: concatErrorCode(ErrorCodePrefix.Chain, 2),
  }),
  NO_BLOCK_FALLBACK: (): Error => ({
    message:
      'could not fall back to pubhashblock ZMQ filter because it is not enabled',
    code: concatErrorCode(ErrorCodePrefix.Chain, 3),
  }),
  INVALID_COOKIE_FILE: (path: string): Error => ({
    message: `invalid cookie authentication file: ${path}`,
    code: concatErrorCode(ErrorCodePrefix.Chain, 4),
  }),
  NO_AUTHENTICATION: (): Error => ({
    message: 'no or invalid authentication specified',
    code: concatErrorCode(ErrorCodePrefix.Chain, 5),
  }),
  ZMQ_CONNECTION_TIMEOUT: (
    symbol: string,
    filter: string,
    address: string,
  ): Error => ({
    message: `connection attempt to ${symbol} ZMQ filter ${filter} (${address}) timed out`,
    code: concatErrorCode(ErrorCodePrefix.Chain, 6),
  }),
};
