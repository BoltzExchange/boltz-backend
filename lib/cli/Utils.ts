import { Arguments } from 'yargs';
import { address, ECPair, Transaction } from 'bitcoinjs-lib';
import { Networks, detectSwap, constructClaimTransaction, constructRefundTransaction } from 'boltz-core';
import inquire from './inquire';
import { getHexBuffer } from '../Utils';
import { OutputType, OrderSide } from '../proto/boltzrpc_pb';

export const getOrderSide = (side: string) => {
  switch (side.toLowerCase()) {
    case 'buy': return OrderSide.BUY;
    case 'sell': return OrderSide.SELL;

    default: throw `could not find order side: ${side}`;
  }
};

export const getOutputType = (type: string) => {
  switch (type.toLowerCase()) {
    case 'bech32': return OutputType.BECH32;
    case 'compatibility': return OutputType.COMPATIBILITY;

    default: return OutputType.LEGACY;
  }
};

export const getNetwork = (networkKey: string) => {
  const network = Networks[networkKey];

  if (!network) {
    throw `Could not find network: ${networkKey}`;
  }

  return network;
};

const parseParameters = (argv: Arguments) => {
  const network = getNetwork(argv.network);

  return {
    network,
    lockupTransaction: Transaction.fromHex(argv.lockup_transaction),
    redeemScript: getHexBuffer(argv.redeem_script),
    destinationScript: address.toOutputScript(argv.destination_address, network),
  };
};

const parseSwapOutput = (redeemScript: Buffer, lockupTransaction: Transaction) => {
  const swapOutput = detectSwap(redeemScript, lockupTransaction);

  if (!swapOutput) {
    throw 'Could not find swap in transaction';
  }

  return swapOutput;
};

export const parseCommands = async (inquiries: any[], argv: Arguments): Promise<Arguments> => {
  const argvLength = Object.keys(argv).length;
  if (argvLength === inquiries.length) {
    const answers = await inquire(inquiries);
    return { ...answers, ...argv };
  } else {
    return argv;
  }
};

export const claimSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const claimKeys = ECPair.fromPrivateKey(getHexBuffer(argv.claim_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const claimTransaction = constructClaimTransaction(
    {
      redeemScript,
      preimage: getHexBuffer(argv.preimage),
      keys: claimKeys,
    },
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return claimTransaction.toHex();
};

export const refundSwap = (argv: Arguments) => {
  const { network, lockupTransaction, redeemScript, destinationScript } = parseParameters(argv);
  const refundKeys = ECPair.fromPrivateKey(getHexBuffer(argv.refund_private_key), { network });

  const swapOutput = parseSwapOutput(redeemScript, lockupTransaction);

  const refundTransaction = constructRefundTransaction(
    refundKeys,
    redeemScript,
    argv.timeout_block_height,
    {
      txHash: lockupTransaction.getHash(),
      ...swapOutput,
    },
    destinationScript,
    argv.fee_per_byte,
  );

  return refundTransaction.toHex();
};
