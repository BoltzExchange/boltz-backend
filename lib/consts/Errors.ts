import { concatErrorCode } from '../Utils';
import { ErrorCodePrefix } from './Enums';
import type { Error } from './Types';

export default {
  IS_DISCONNECTED: (clientName: string): Error => ({
    message: `${clientName} is disconnected`,
    code: concatErrorCode(ErrorCodePrefix.General, 1),
  }),
  COULD_NOT_PARSE_CONFIG: (filename: string, error: string): Error => ({
    message: `could not parse ${filename} config: ${error}`,
    code: concatErrorCode(ErrorCodePrefix.General, 2),
  }),
};
