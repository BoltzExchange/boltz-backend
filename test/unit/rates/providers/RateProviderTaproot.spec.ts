import { getPairId, hashString } from '../../../../lib/Utils';
import { OrderSide, SwapVersion } from '../../../../lib/consts/Enums';
import FeeProvider from '../../../../lib/rates/FeeProvider';
import RateProviderTaproot from '../../../../lib/rates/providers/RateProviderTaproot';
import Errors from '../../../../lib/service/Errors';

jest.mock('../../../../lib/rates/FeeProvider', () => {
  return jest.fn().mockImplementation(() => {
    return {
      minerFees: new Map<string, any>([
        [
          'BTC',
          {
            [SwapVersion.Taproot]: {
              normal: 151,
              reverse: {
                lockup: 154,
                claim: 111,
              },
            },
          },
        ],
        [
          'L-BTC',
          {
            [SwapVersion.Taproot]: {
              normal: 1337,
              reverse: {
                lockup: 2503,
                claim: 1297,
              },
            },
          },
        ],
      ]),
      getPercentageFees: jest.fn().mockReturnValue({
        percentage: 0.5,
        percentageSwapIn: 0.1,
      }),
      getBaseFee: jest.fn().mockImplementation((currency: string) => {
        switch (currency) {
          case 'BTC':
            return 12123;

          default:
            return 252;
        }
      }),
    };
  });
});

const mockedFeeProvider = (<jest.Mock<FeeProvider>>(<any>FeeProvider))();

describe('RateProviderTaproot', () => {
  const pairConfigs = [
    {
      base: 'L-BTC',
      quote: 'BTC',
      rate: 1,
      minSwapAmount: 1_000,
      maxSwapAmount: 1_000_000,
    },
  ];

  const provider = new RateProviderTaproot(
    new Map<string, any>([
      ['BTC', { chainClient: {}, lndClient: {} }],
      [
        'L-BTC',
        {
          chainClient: {},
        },
      ],
      ['R-BTC', { provider: {} }],
      ['CASHU', {}],
    ]),
    mockedFeeProvider,
    new Map<string, any>(pairConfigs.map((pair) => [getPairId(pair), pair])),
    new Map<string, number>([
      ['BTC', 0],
      ['L-BTC', 100_000],
    ]),
  );

  beforeEach(() => {
    provider.reversePairs.clear();
    provider.submarinePairs.clear();

    jest.clearAllMocks();
  });

  test('should serialize pairs', () => {
    provider.submarinePairs.set(
      'BTC',
      new Map<string, any>([
        [
          'L-BTC',
          {
            hash: 'data',
          },
        ],
      ]),
    );
    provider.submarinePairs.set('deleted', new Map());

    const obj = RateProviderTaproot.serializePairs(provider.submarinePairs);
    expect(obj).toEqual({
      BTC: {
        'L-BTC': {
          hash: 'data',
        },
      },
    });
  });

  test('should set hardcoded pair', () => {
    const rate = 21;

    provider.setHardcodedPair({
      rate,
      base: 'L-BTC',
      quote: 'BTC',
    } as any);

    expect(provider.submarinePairs.get('L-BTC')!.get('BTC')!.rate).toEqual(
      1 / rate,
    );
    expect(provider.reversePairs.get('BTC')!.get('L-BTC')!.rate).toEqual(rate);
  });

  test('should update pair', () => {
    const rate = 21;

    provider.updatePair('L-BTC/BTC', rate);

    expect(provider.submarinePairs.get('L-BTC')!.get('BTC')!.rate).toEqual(
      1 / rate,
    );
    expect(provider.reversePairs.get('BTC')!.get('L-BTC')!.rate).toEqual(rate);
  });

  test('should update hardcoded pairs', () => {
    provider.setHardcodedPair({
      rate: 1,
      base: 'L-BTC',
      quote: 'BTC',
    } as any);

    const wrongHash = 'wrongHash';

    provider.submarinePairs.get('L-BTC')!.get('BTC')!.hash = wrongHash;
    provider.reversePairs.get('BTC')!.get('L-BTC')!.hash = wrongHash;

    provider.updateHardcodedPair('L-BTC/BTC');

    expect(provider.submarinePairs.get('L-BTC')!.get('BTC')!.hash).not.toEqual(
      wrongHash,
    );
    expect(provider.reversePairs.get('BTC')!.get('L-BTC')!.hash).not.toEqual(
      wrongHash,
    );
  });

  test('should validate pair hash', () => {
    const hash = 'hash';

    provider.submarinePairs.set(
      'BTC',
      new Map<string, any>([
        [
          'L-BTC',
          {
            hash,
          },
        ],
      ]),
    );

    provider.validatePairHash(hash, 'L-BTC/BTC', OrderSide.BUY, false);
  });

  test('should throw when validating pair hash when hash is invalid', () => {
    provider.submarinePairs.set(
      'BTC',
      new Map<string, any>([
        [
          'L-BTC',
          {
            hash: 'noMatch',
          },
        ],
      ]),
    );

    expect(() =>
      provider.validatePairHash('hash', 'L-BTC/BTC', OrderSide.BUY, false),
    ).toThrow(Errors.INVALID_PAIR_HASH().message);
  });

  test('should throw when validating pair hash of pair that does not exit', () => {
    const pair = 'not/found';
    expect(() =>
      provider.validatePairHash('hash', pair, OrderSide.BUY, false),
    ).toThrow(Errors.PAIR_NOT_FOUND(pair).message);

    provider.submarinePairs.set('found', new Map());

    expect(() =>
      provider.validatePairHash('hash', pair, OrderSide.BUY, false),
    ).toThrow(Errors.PAIR_NOT_FOUND(pair).message);
  });

  test('should getNested pair map', () => {
    const nested = provider['getToMap'](
      'not/found',
      OrderSide.BUY,
      false,
      true,
    )!;
    nested.toMap.set('yo', { test: 'data' } as any);

    expect(nested).toEqual({
      fromAsset: 'found',
      toAsset: 'not',
      toMap: expect.any(Map),
    });

    expect(
      provider['getToMap']('not/found', OrderSide.BUY, false, false),
    ).toEqual({
      fromAsset: 'found',
      toAsset: 'not',
      toMap: nested.toMap,
    });

    expect(provider['getToMap']('not/found', OrderSide.BUY, true)).toEqual(
      undefined,
    );
  });

  test('should hash pairs', () => {
    const pair = {
      hash: 'taproot',
      rate: 21,
      fees: {
        percentage: 1,
        minerFees: 321,
      },
      limits: {
        minimal: 98,
        maximal: 890,
      },
    } as any;

    expect(provider['hashPair'](pair)).toEqual(
      hashString(
        JSON.stringify({
          rate: pair.rate,
          fees: pair.fees,
          limits: pair.limits,
        }),
      ),
    );
    expect(provider['hashPair'](pair)).toMatchSnapshot();

    expect(provider['hashPair']({ ...pair, other: 'data' })).toEqual(
      provider['hashPair'](pair),
    );
  });

  test('should set pair with rate', () => {
    provider['setPair']('L-BTC/BTC', OrderSide.BUY, true, 1, {
      miner: 'fees',
    } as any);

    expect(provider.reversePairs.size).toEqual(1);
    expect(provider.reversePairs.get('BTC')!.size).toEqual(1);
    expect(provider.reversePairs.get('BTC')!.get('L-BTC')).toEqual({
      hash: expect.anything(),
      rate: 1,
      limits: {
        maximal: 1_000_000,
        minimal: 1_000,
      },
      fees: {
        percentage: 0.5,
        minerFees: {
          miner: 'fees',
        },
      },
    });

    expect(provider.submarinePairs.size).toEqual(0);
  });

  test('should set pair and get miner fees when not provided', () => {
    provider['setPair']('L-BTC/BTC', OrderSide.BUY, true, 1);

    expect(provider.reversePairs.size).toEqual(1);
    expect(provider.reversePairs.get('BTC')!.size).toEqual(1);
    expect(provider.reversePairs.get('BTC')!.get('L-BTC')).toEqual({
      hash: expect.anything(),
      rate: 1,
      limits: {
        maximal: 1_000_000,
        minimal: 1_000,
      },
      fees: {
        percentage: 0.5,
        minerFees: {
          claim: 1297,
          lockup: 2503,
        },
      },
    });

    expect(provider.submarinePairs.size).toEqual(0);
  });

  test('should not set pair no rate is provided and none was set prior', () => {
    provider['setPair']('L-BTC/BTC', OrderSide.BUY, true);

    expect(provider.reversePairs.size).toEqual(1);
    expect(provider.reversePairs.get('BTC')!.size).toEqual(0);
    expect(provider.submarinePairs.size).toEqual(0);
  });

  test('should not set pair when combination is not possible', () => {
    provider['setPair']('L-BTC/BTC', OrderSide.BUY, false);
    expect(provider.reversePairs.size).toEqual(0);
    expect(provider.submarinePairs.size).toEqual(1);
    expect(provider.submarinePairs.get('BTC')!.size).toEqual(0);
  });

  test('should get limits with maximal 0-conf amount for reverse swaps', () => {
    expect(provider['getLimits']('L-BTC/BTC', OrderSide.BUY, true, 1)).toEqual({
      minimal: 1_000,
      maximal: 1_000_000,
    });
  });

  test.each`
    side              | maximalZeroConf
    ${OrderSide.BUY}  | ${0}
    ${OrderSide.SELL} | ${100_000}
  `(
    'should get limits with maximal 0-conf amount for submarine swaps',
    ({ side, maximalZeroConf }) => {
      expect(provider['getLimits']('L-BTC/BTC', side, false, 1)).toEqual({
        maximalZeroConf,
        minimal: 1_000,
        maximal: 1_000_000,
      });
    },
  );

  test('should throw when getting limits for pair that does not exist', () => {
    const pair = 'not/found';
    expect(() => provider['getLimits'](pair, OrderSide.BUY, false, 1)).toThrow(
      `Could not get limits for pair: ${pair}`,
    );
  });

  test.each`
    isReverse | from       | to         | expected
    ${true}   | ${'BTC'}   | ${'BTC'}   | ${true}
    ${false}  | ${'BTC'}   | ${'BTC'}   | ${true}
    ${true}   | ${'BTC'}   | ${'L-BTC'} | ${true}
    ${false}  | ${'BTC'}   | ${'L-BTC'} | ${false}
    ${true}   | ${'L-BTC'} | ${'BTC'}   | ${false}
    ${false}  | ${'L-BTC'} | ${'BTC'}   | ${true}
  `(
    'should check if is possible combination',
    ({ isReverse, from, to, expected }) => {
      expect(provider['isPossibleCombination'](isReverse, from, to)).toEqual(
        expected,
      );
    },
  );

  test.each`
    currency      | expected
    ${'BTC'}      | ${true}
    ${'L-BTC'}    | ${false}
    ${'notFound'} | ${false}
  `(
    'should check if $currency can be used on lightning',
    ({ currency, expected }) => {
      expect(provider['canLightning'](currency)).toEqual(expected);
    },
  );

  test.each`
    currency   | expected
    ${'BTC'}   | ${true}
    ${'L-BTC'} | ${true}
    ${'R-BTC'} | ${true}
    ${'CASHU'} | ${false}
  `(
    'should check if $currency can be used onchain',
    ({ currency, expected }) => {
      expect(provider['canOnchain'](currency)).toEqual(expected);
    },
  );
});
