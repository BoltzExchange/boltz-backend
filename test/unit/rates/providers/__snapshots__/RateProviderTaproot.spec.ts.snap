// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RateProviderTaproot should hash pairs 1`] = `"aa172104157f71f6a7edda34d6551f3a01e4d39477dae0956e9405f57d9d6acc"`;
